### PR TITLE
test: red tests for shell parsing bypass bugs

### PR DESCRIPTION
**Pull request by cameronsjo**
## What

7 failing (red) tests documenting confirmed shell parsing security gaps.

### Redirect parser: no ANSI-C (`$'...'`) state

- `clobber_redirect_ansi_c_escaped_quote_should_not_bypass` — `echo $'a\'b' > .env`
  The redirect parser has `'` and `"` tracking but not `$'...'`.  The escaped
  quote (`\'`) looks like a close, the real `'` reopens a phantom string,
  and the `>` after it is swallowed. Target (`.env`) never seen.

- `redirect_append_ansi_c_escaped_quote_should_not_bypass` — same via `>> .env`

### `$()` early close + unmatched quote

3 e2e guard tests (`prevent_secret_leaks`):

- `subst_early_close_single_quote_should_not_bypass` — `echo $(echo ') && cat .env`
  The `)` inside the single-quoted string closes `$()` early (no depth tracking
  in `substitution_bodies`), the unmatched `'` at position 15 swallows everything
  after it including `&& cat .env`. Outer segment head is `echo` (metadata-safe)
  → never checked.

- `subst_early_close_double_quote_should_not_bypass` — same with `"a)`

- `subst_early_close_with_different_reader_should_not_bypass` — same with
  `grep foo $(echo ') && cat .env` (non-echo head)

### Segmenter unit tests (green, for contrast)

2 `split_segments` unit tests that the parser _does_ split correctly on these
inputs — the bug is that the guard chain doesn't check all resulting segments.

## Why

These are real bypasses: valid bash commands that would read/write secrets are
completely invisible to the guard. The `$()` early-close + unmatched-quote
pattern is sharpest — a dangerous tail command is swallowed by the dangling
quote and never reaches any checker.

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened protection against shell commands that could bypass secret-file read restrictions through complex command substitutions and unmatched quotes.
  * Improved detection of clobber and append redirects following ANSI-C quoted strings with escaped quotes, helping prevent unauthorized secret-file writes.

* **Tests**
  * Added regression coverage for command parsing, secret reads, and redirect protection in these edge cases.

---

**Fix landed** (Fable orchestration, Opus implementation + security adjudication): two shell-parser desyncs closed in `crates/core/src/shell.rs`. All 7 red tests green; Opus security review SHIP (33-case differential: 14 closed misses, 1 removed false-block, no missed command). Four pre-existing/bounded follow-ups filed: #652, #653, #654, #655. Releases 0.75.0.

```
Session-Name: pr-slates-2026-08-08
Session-Id: d7cca765-6b18-4f59-a720-06996dd637b1
Model: claude-fable-5
Harness: claude-code 2.1.226
Machine: cf6e768835c7
```

### CODE REVIEW

**Comment by cameronsjo** [comment]
### The 7 / 9 / 5 discrepancy is a `cargo test` artifact, not a test-count disagreement

Three numbers were in play — the body says 7 red, the diff carries 9 `#[test]`
functions, and CI reports 5 failing. All three are correct and consistent.

`cargo test --workspace` (what `Makefile:test` and `ci.yml` both run) builds and
runs test binaries crate by crate, and **stops after the first binary that
fails** unless `--no-fail-fast` is passed. The `cadence-hooks-cadence` lib
binary fails first with 5 failures and aborts the run — `cadence-hooks-core`'s
binary is never invoked at all. `Running unittests ... cadence_hooks_core-*`
appears nowhere in the CI log, which is the tell.

Running that crate in isolation:

```
$ cargo test -p cadence-hooks-core --lib
test result: FAILED. 716 passed; 2 failed
```

5 + 2 = **7 red**, matching the body. The remaining 2 are the
`split_segments_*` green-for-contrast pair, both confirmed passing. 7 + 2 = 9.

Ruled out along the way: none of the 9 tests exist on `main`, so "a test started
passing after an unrelated fix" cannot explain the gap. The one commit touching
these files between the branch point and current `main` (#556, git-push option
grammar) only adds functions — it modifies none of the parsing functions under
test.

**Implication for CI:** while this branch is red, any *other* failure in
`cadence-hooks-core` is invisible on it — the binary never runs. Worth
`--no-fail-fast` in the workflow regardless of what happens to this PR.

**One caveat on the green pair:** one of the two passes for a weaker reason than
its name implies — its assertion does not distinguish the correct result from
the buggy one. Flagging it so it is not read as coverage when the fix lands;
details separately rather than here.

File: crates/core/src/shell.rs [hunk 1]
```diff
@@ -4873,6 +4873,31 @@ mod tests {
...
+    fn split_segments_subst_early_close_double_quote_should_not_swallow_rest() {
+        // Same bug with a double quote. The `)` inside `"…"` closes the
+        // `$()` early (no depth tracking), the `"` stays open and swallows
+        // the rest of the line.
+        let out = split_segments(r#"echo $(echo "a) && cat .env"#);
+        assert!(
+            out.iter().any(|s| s.contains("cat .env")),
+            "unmatched double-quote swallowed the second command: {out:?}"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/shell.rs:5198]: Declining. The concatenation convention applies to the installed **terminology guard's** blocklist words (which it blocks in any Write to a non-excluded path), not to `.env` as a string literal in Rust test source. These fixtures must carry the literal `.env` precisely because they exercise `prevent-secret-writes` — composing the target filename from fragments would obscure what each case tests for no safety gain, and `crates/core/` source content is not inspected by the redirect/secret-write guard (only Bash command strings and Write `file_path`s are).
